### ISSUE TITLE
Command line option to load extra config file

### ISSUE THREAD

**Issue by tamlyn** [lorenwest/node-config#249]
I know I can load extra config *values* with the `NODE_CONFIG` command line option but I would like the option of loading values from a config *file*. For example:

   $ node myapp.js --NODE_CONFIG_FILE=testing

Would load `config/testing.{EXT}`.

My use case is that I want to set some options when running tests (e.g. raise log level to avoid polluting test output, use a different DB name for integration tests). I also want to be able to run the tests in different environments (e.g. with caching turned on or off) and on different hosts (e.g. with different DB connection options). Finally, I would prefer to have as much config as possible in files in git rather than spread out in environment variables and command line switches.

I'm happy to have a go at implementing this feature but I wanted to check whether it's something that would be considered or whether there's another way of achieving what I want.

**Comment by lorenwest** [comment]
Hi @tamlyn,  

The wiki page on [config files](https://github.com/lorenwest/node-config/wiki/Configuration-Files) should answer your questions about saving configs in files and having different configs for different test/deploy environments.

Close out the issue if that answers your questions.

Cheers,
-Loren

**Comment by tamlyn** [comment]
My problem is that I wan to apply a set of configs to an existing `default` + `deployment` + `host` config set. The on;y way I've found to do it (and keep the config in a file) is to hijack the `NODE_APP_INSTANCE` environment variable and set it to `testing` which then loads `default-testing.js` which I can apply to either my `development` or `production` deployment configs. Does that make sense?

**Comment by markstos** [comment]
@tamlyn I wouldn't call that hijacking `NODE_APP_INSTANCE`. I think that's a common use of the variable and how I use it myself. What are doing now is the solution I would recommend.

**Comment by tamlyn** [comment]
OK, thanks. Still seems weird to me and not obvious for a new developer.

**Comment by tamlyn** [comment]
I'm running into issues with my above solution. For instance I want to have different logging settings on `production` and `development` environments and no logging when running tests (which may run in either environment). Using `NODE_APP_INSTANCE` I would need to have to duplicate my config in `development-testing.js` and `production-testing.js` otherwise the settings in `development.js` and `production.js` override the settings in `default-testing.js`.

I've considered instructing gulp to load the testing config from another file and serialise it to `NODE_CONFIG` env but that makes it difficult to run the tests outside of gulp (when debugging, for instance). And it's also weird for anyone trying to work out what's going on.

Being able to specify an extra file to load as a command line option would really help and seems not that different from specifying config as JSON. In fact we could use the same `NODE_CONFIG` parameter and auto-detect whether it is JSON or a simple string to use as a file name.

**Comment by markstos** [comment]
@tamlyn would this solution work for you? 

Use `custom-environment-variables.json` to map DOES_LOGGING to your `does_logging` config key. 

Then in your `gulpfile.js` or equivalent, update the `test` command to set `DOES_LOGGING=1`. 

Or for that matter, the config object remains mutable until the first call to `get()` to allow for exceptional cases like this.   After loading `config` in your `gulpfile` before running the tests, you could set `config.does_logging = 1`.

**Comment by tamlyn** [comment]
The thing is there are a bunch of config options that are different when running tests (e.g. don't send emails, use a different redis db...). I also want to be able to run the tests without gulp (e.g. for debugging with node-inspector) so using lots of environment variables would be a pain.

In the end I've switched back to `NODE_ENV=testing` and I've moved my configs which previously differed between `development` and `production` into host-specific configs. It works and it's probably more how it's intended to be used.

Thanks for the help.
